File: frontend/lib/models/ai.dart
/// Models for the AI study-plan feature.

/// Non-secret view of the user's AI settings returned by the backend.
class AiSettings {
  final bool configured;
  final String? provider;
  final String? model;
  final String? apiKeyPreview;

  const AiSettings({
    required this.configured,
    this.provider,
    this.model,
    this.apiKeyPreview,
  });

  factory AiSettings.fromJson(Map<String, dynamic> json) => AiSettings(
        configured: json['configured'] as bool? ?? false,
        provider: json['provider'] as String?,
        model: json['model'] as String?,
        apiKeyPreview: json['api_key_preview'] as String?,
      );
}

/// A single chat turn. `role` is 'user' or 'assistant'.
class ChatMessage {
  final String role;
  final String content;

  const ChatMessage({required this.role, required this.content});

  bool get isUser => role == 'user';

  Map<String, dynamic> toJson() => {'role': role, 'content': content};
}

/// One day of an AI-generated plan (preview before import).
class GeneratedDay {
  final int day;
  final int? week;
  final String? phase;
  final String? focus;
  final String? task;
  final String? difficulty;
  final String? notes;

  const GeneratedDay({
    required this.day,
    this.week,
    this.phase,
    this.focus,
    this.task,
    this.difficulty,
    this.notes,
  });

  factory GeneratedDay.fromJson(Map<String, dynamic> json) => GeneratedDay(
        day: json['day'] as int,
        week: json['week'] as int?,
        phase: json['phase'] as String?,
        focus: json['focus'] as String?,
        task: json['task'] as String?,
        difficulty: json['difficulty'] as String?,
        notes: json['notes'] as String?,
      );

  Map<String, dynamic> toJson() => {
        'day': day,
        if (week != null) 'week': week,
        if (phase != null) 'phase': phase,
        if (focus != null) 'focus': focus,
        if (task != null) 'task': task,
        if (difficulty != null) 'difficulty': difficulty,
        if (notes != null) 'notes': notes,
      };
}

/// The full structured plan the AI produced.
class GeneratedPlan {
  final String name;
  final String? description;
  final int totalDays;
  final List<GeneratedDay> days;

  const GeneratedPlan({
    required this.name,
    this.description,
    required this.totalDays,
    required this.days,
  });

  factory GeneratedPlan.fromJson(Map<String, dynamic> json) => GeneratedPlan(
        name: json['name'] as String? ?? 'AI Study Plan',
        description: json['description'] as String?,
        totalDays: json['total_days'] as int? ?? 0,
        days: ((json['days'] as List?) ?? [])
            .map((e) => GeneratedDay.fromJson((e as Map).cast<String, dynamic>()))
            .toList(),
      );

  Map<String, dynamic> toJson() => {
        'name': name,
        if (description != null) 'description': description,
        'total_days': totalDays,
        'days': days.map((d) => d.toJson()).toList(),
      };

  /// Distinct week count for a compact preview summary.
  int get weekCount => days.map((d) => d.week).whereType<int>().toSet().length;
}

File: backend/app/schemas/ai.py
"""Schemas for the AI plan-generation feature."""

from __future__ import annotations

from datetime import date

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.plan import ImportSummary


# -- Settings ---------------------------------------------------------------
class AISettingsUpdate(BaseModel):
    """Payload to save/replace the user's provider API key."""

    api_key: str = Field(..., min_length=8, max_length=400)
    provider: str = Field(default="nvidia", max_length=50)
    model: str | None = Field(default=None, max_length=150)


class AISettingsResponse(BaseModel):
    """Non-secret view of the user's AI settings (never returns the key)."""

    model_config = ConfigDict(from_attributes=True)

    configured: bool
    provider: str | None = None
    model: str | None = None
    api_key_preview: str | None = Field(
        default=None, description="Masked preview, e.g. 'nvapi-****a1b2'."
    )


# -- Chat -------------------------------------------------------------------
class ChatMessage(BaseModel):
    role: str = Field(..., pattern="^(user|assistant)$")
    content: str = Field(..., min_length=1, max_length=8000)


class ChatRequest(BaseModel):
    """A user turn plus prior conversation for context."""

    message: str = Field(..., min_length=1, max_length=8000)
    history: list[ChatMessage] = Field(default_factory=list, max_length=20)


class ChatResponse(BaseModel):
    reply: str
    model: str


# -- Structured plan generation --------------------------------------------
class GeneratePlanRequest(BaseModel):
    """Ask the AI to convert the agreed-upon plan into structured rows.

    ``conversation`` is the full chat so far so the model has the plan it just
    proposed in context.
    """

    conversation: list[ChatMessage] = Field(..., min_length=1, max_length=30)
    instruction: str | None = Field(
        default=None,
        max_length=2000,
        description="Optional final tweak, e.g. 'make it 8 weeks'.",
    )


class GeneratedDay(BaseModel):
    day: int
    week: int | None = None
    phase: str | None = None
    focus: str | None = None
    task: str | None = None
    difficulty: str | None = None
    notes: str | None = None


class GeneratedPlan(BaseModel):
    """The AI's structured plan, previewed before the user confirms import."""

    name: str
    description: str | None = None
    total_days: int
    days: list[GeneratedDay]


# -- Confirm & import -------------------------------------------------------
class ConfirmImportRequest(BaseModel):
    """User confirms a previously-generated plan and imports it."""

    plan: GeneratedPlan
    start_date: date = Field(..., description="Date mapped to Day 1.")


class ConfirmImportResponse(BaseModel):
    summary: ImportSummary
